packages/babel-helpers/src/helpers/asyncToGenerator.ts

```diff
@@ -10,22 +18,22 @@ function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
   }
 
   if (info.done) {
-    resolve(value);
+    resolve(value as T);
```

Comment:
Is this cast correct? If `info.done` the the generator is returning, it's not yielding.

Maybe `asyncGeneratorStep` should accept `<T, R>`, gen should be `Generator<T, R, unknown>`, and then TS will auto-infer this as `R`?

Reply:
Oh TS will not auto-infer this as `R` because we are using `var` and not `const` (so it looses track of the relation between `info`, `info.done`, and `value`. However, I still think it should be `R`.

Reply:
Ok, after playing around in the TS playground I think these would be the correct signatures:
```ts
function asyncGeneratorStep<T, R>(
  gen: Generator<T, R, Awaited<T>>,
  resolve: (value: R) => void,
  reject: (error: any) => void,
  _next: (value: Awaited<T> | undefined) => void,
  _throw: (err: any) => void,
  key: "next" | "throw",
  arg: Awaited<T> | undefined,
): void;

function _asyncToGenerator<This, Args extends unknown[], R, T>(
  fn: (this: This, ...args: Args) => Generator<T, R, Awaited<T>>,
): (this: This, ...args: Args) => Promise<R>;
```

where `T` is what you pass to `await` in an async function, and `R` is what you use in `return` statements, `This` is the this value of the async function, and `Args` are its arguments.

Reply:
Fixed in [`a9b107d` (#16510)](https://github.com/babel/babel/pull/16510/commits/a9b107d2d66f08d23629a052b99278ccdcec2aba). And I slightly improve the type of `asyncGeneratorStep` a little bit more, using function overload to get the type of the `arg` right.